distribution/pull_v2.go

```diff
@@ -877,6 +886,32 @@ func (p *v2Puller) pullSchema2Config(ctx context.Context, dgst digest.Digest) (c
 	return configJSON, nil
 }
 
+func retry(ctx context.Context, maxAttempts int, sleep time.Duration, f func(ctx context.Context) error) (err error) {
+	attempt := 0
+	for ; attempt < maxAttempts; attempt++ {
+		err = retryOnError(f(ctx))
+		if err == nil {
+			return nil
+		}
+		if xfer.IsDoNotRetryError(err) {
+			break
+		}
+
+		if attempt+1 < maxAttempts {
+			timer := time.NewTicker(sleep)
+			select {
+			case <-ctx.Done():
+				timer.Stop()
+				return ctx.Err()
+			case <-timer.C:
+				logrus.WithError(err).WithField("attempts", attempt+1).Debug("retrying after error")
+				sleep *= 2
```

Comment:
Ah, that would be me to blame. I always confuse them (but noticed that the `xfer` was using a ticker, and I recalled some oddities when to do / don't use them https://github.com/moby/moby/pull/43291#discussion_r813871236

Reply:
Ah, I guess there it's using it for the retrying in XX seconds loop, so it's slightly different; in this case it would probably need to take the whole loop here 🤔

Reply:
Another, more compact approach (actually using a Timer):
```golang
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(sleepDuration):
				break
			}
```

Reply:
I did a quick search for the issue I remembered; and I think that issue was when using `time.After()` see https://github.com/moby/moby/pull/38522

Reply:
I've changed it back to `time.NewTimer()`. I think using time.After() makes a leak if the other select succeeds (in the cast the context times out).

Reply:
Yes, that makes sense, thanks!